drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1615 @@
+  }
+  local credData = command.args.credentialData
+
+  -- Check busy state
+  local busy = device:get_field(lock_utils.BUSY_STATE)
+  if busy == true then
+    local result = {
+      commandName = cmdName,
+      statusCode = "busy"
+    }
+    local event = capabilities.lockCredentials.commandResult(
+      result,
+      {
+        state_change = true,
+        visibility = {displayed = false}
+      }
+    )
+    device:emit_event(event)
+    return
+  end
+
+  -- Save values to field
+  device:set_field(lock_utils.BUSY_STATE, true, {persist = true})
+  device:set_field(lock_utils.COMMAND_NAME, cmdName, {persist = true})
+  device:set_field(lock_utils.USER_INDEX, userIdx, {persist = true})
+  device:set_field(lock_utils.CRED_INDEX, credIdx, {persist = true})
+
+  -- Send command
+  local ep = device:component_to_endpoint(command.component)
+  device:send(
+    DoorLock.server.commands.SetCredential(
+      device, ep,
+      DoorLock.types.DlDataOperationType.MODIFY, -- Data Operation Type: Add(0), Modify(2)
+      credential,  -- Credential
+      credData,    -- Credential Data
+      userIdx,     -- User Index
+      nil,         -- User Status
+      nil          -- User Type
+    )
+  )
+end
+
+-----------------------------
+-- Set Credential Response --
+-----------------------------
+local function set_credential_response_handler(driver, device, ib, response)
+  if ib.status ~= im.InteractionResponse.Status.SUCCESS then
+    device.log.error("Failed to set credential for device")
+    return
+  end
+
+  local cmdName = device:get_field(lock_utils.COMMAND_NAME)
+  local userIdx = device:get_field(lock_utils.USER_INDEX)
+  local credIdx = device:get_field(lock_utils.CRED_INDEX)
+  local status = "success"
+  local elements = ib.info_block.data.elements
+  if elements.status.value == DoorLock.types.DlStatus.SUCCESS then
+    -- If user is added also, update User table
+    if userIdx == nil then
+      local userType = device:get_field(lock_utils.USER_TYPE)
+      add_user_to_table(device, elements.user_index.value, userType)
+    end
+
+    -- Update Credential table
+    userIdx = elements.user_index.value
+    if cmdName == "addCredential" then
+      add_credential_to_table(device, userIdx, credIdx, "pin")
+    end
+
+    -- Update commandResult
+    local result = {
+      commandName = cmdName,
+      userIndex = userIdx,
+      credentialIndex = credIdx,
+      statusCode = status
+    }
+    local event = capabilities.lockCredentials.commandResult(
+      result,
+      {
+        state_change = true,
+        visibility = {displayed = false}
+      }
+    )
+    device:emit_event(event)
+    device:set_field(lock_utils.BUSY_STATE, false, {persist = true})
+    return
+  end
+
+  -- Update commandResult
+  status = "occupied"
+  if elements.status.value == DoorLock.types.DlStatus.FAILURE then
+    status = "failure"
+  elseif elements.status.value == DoorLock.types.DlStatus.DUPLICATE then
+    status = "duplicate"
```

Comment:
We should add handling for `duplicate` codes, at least for COTA PINs. If the lock returns the duplicate error code we should create a new randomized COTA PIN and use that instead. I think it's OK if that doesn't land in this initial PR but if it doesn't please create a ticket to follow up on it.